src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.ReadWrite.cs

```diff
@@ -0,0 +1,431 @@
+                }
+
+                if (buffer.IsSingleSegment)
+                {
+                    var pDataChunks = stackalloc HttpApiTypes.HTTP_DATA_CHUNK[1];
+
+                    fixed (byte* pBuffer = &MemoryMarshal.GetReference(buffer.First.Span))
+                    {
+                        ref var chunk = ref pDataChunks[0];
+
+                        chunk.DataChunkType = HttpApiTypes.HTTP_DATA_CHUNK_TYPE.HttpDataChunkFromMemory;
+                        chunk.fromMemory.pBuffer = (IntPtr)pBuffer;
+                        chunk.fromMemory.BufferLength = (uint)buffer.Length;
+                        hr = NativeMethods.http_write_response_bytes(_pInProcessHandler, pDataChunks, nChunks, out fCompletionExpected);
+                    }
+                }
+                else
+                {
+                    // REVIEW: Do we need to guard against this getting too big? It seems unlikely that we'd have more than say 10 chunks in real life
+                    var pDataChunks = stackalloc HttpApiTypes.HTTP_DATA_CHUNK[nChunks];
+                    var currentChunk = 0;
+
+                    // REVIEW: We don't really need this list since the memory is already pinned with the default pool,
+                    // but shouldn't assume the pool implementation right now. Unfortunately, this causes a heap allocation...
+                    var handles = new MemoryHandle[nChunks];
+
+                    foreach (var b in buffer)
+                    {
+                        ref var handle = ref handles[currentChunk];
+                        ref var chunk = ref pDataChunks[currentChunk];
+
+                        handle = b.Retain(true);
+
+                        chunk.DataChunkType = HttpApiTypes.HTTP_DATA_CHUNK_TYPE.HttpDataChunkFromMemory;
+                        chunk.fromMemory.BufferLength = (uint)b.Length;
+                        chunk.fromMemory.pBuffer = (IntPtr)handle.Pointer;
+
+                        currentChunk++;
+                    }
+
+                    hr = NativeMethods.http_write_response_bytes(_pInProcessHandler, pDataChunks, nChunks, out fCompletionExpected);
+                    // Free the handles
+                    foreach (var handle in handles)
+                    {
+                        handle.Dispose();
+                    }
+                }
+
+                if (!fCompletionExpected)
+                {
+                    _operation.Complete(hr, 0);
+                }
+                return _operation;
+            }
+        }
+
+        private unsafe IISAwaitable FlushToIISAsync()
+        {
+            // Calls flush 
+            var hr = 0;
+            hr = NativeMethods.http_flush_response_bytes(_pInProcessHandler, out var fCompletionExpected);
+            if (!fCompletionExpected)
+            {
+                _operation.Complete(hr, 0);
+            }
+
+            return _operation;
+        }
+
+        public async Task ReadAndWriteLoopAsync()
```

Comment:
One thing that would help grok this would be a summary of the functionality in this method and the various ways each of the state (_doneReading, and _doneWriting) ends up being set.